target/scripts/startup/setup-stack.sh

```diff
@@ -345,28 +346,98 @@ function _setup_dovecot_local_user
         IFS=':' ; read -r -a USER_QUOTA < <(grep "${USER}@${DOMAIN}:" -i /tmp/docker-mailserver/dovecot-quotas.cf)
         unset IFS
 
-        [[ ${#USER_QUOTA[@]} -eq 2 ]] && USER_ATTRIBUTES="${USER_ATTRIBUTES} userdb_quota_rule=*:bytes=${USER_QUOTA[1]}"
+        if [[ ${#USER_QUOTA[@]} -eq 2 ]]
+        then
+          USER_ATTRIBUTES="${USER_ATTRIBUTES:+${USER_ATTRIBUTES} }userdb_quota_rule=*:bytes=${USER_QUOTA[1]}"
+        fi
       fi
 
-      # Let's go!
-      _notify 'inf' "user '${USER}' for domain '${DOMAIN}' with password '********', attr=${USER_ATTRIBUTES}"
+      if [[ -z ${USER_ATTRIBUTES} ]]
+      then
+        _notify 'inf' "Creating user '${USER}' for domain '${DOMAIN}'"
+      else
+        _notify 'inf' "Creating user '${USER}' for domain '${DOMAIN}' with attributes '${USER_ATTRIBUTES}'"
+      fi
 
       echo "${LOGIN} ${DOMAIN}/${USER}/" >> /etc/postfix/vmailbox
-      # User database for dovecot has the following format:
-      # user:password:uid:gid:(gecos):home:(shell):extra_fields
-      # Example :
-      # ${LOGIN}:${PASS}:5000:5000::/var/mail/${DOMAIN}/${USER}::userdb_mail=maildir:/var/mail/${DOMAIN}/${USER}
-      echo "${LOGIN}:${PASS}:5000:5000::/var/mail/${DOMAIN}/${USER}::${USER_ATTRIBUTES}" >> /etc/dovecot/userdb
+      # Dovecot's userdb has the following format
+      #   user:password:uid:gid:(gecos):home:(shell):extra_fields
+      echo \
+        "${LOGIN}:${PASS}:5000:5000::/var/mail/${DOMAIN}/${USER}::${USER_ATTRIBUTES}" \
+        >>/etc/dovecot/userdb
+
       mkdir -p "/var/mail/${DOMAIN}/${USER}"
 
-      # Copy user provided sieve file, if present
+      # copy user provided sieve file, if present
       if [[ -e "/tmp/docker-mailserver/${LOGIN}.dovecot.sieve" ]]
       then
         cp "/tmp/docker-mailserver/${LOGIN}.dovecot.sieve" "/var/mail/${DOMAIN}/${USER}/.dovecot.sieve"
       fi
 
       echo "${DOMAIN}" >> /tmp/vhost.tmp
     done < <(grep -v "^\s*$\|^\s*\#" /tmp/docker-mailserver/postfix-accounts.cf)
+
+    # see https://github.com/docker-mailserver/docker-mailserver/pull/2248#issuecomment-953313852
+    # for more details on this section
+    if [[ -f /tmp/docker-mailserver/postfix-virtual.cf ]] && [[ ${ENABLE_QUOTAS} -eq 1 ]]
+    then
+      # adding aliases to Dovecot's userdb
+      # ${REAL_FQUN} is a user's fully-qualified username
+      local ALIAS REAL_FQUN
+      while read -r ALIAS REAL_FQUN
+      do
+        # ignore comments
+        [[ ${ALIAS} == \#* ]] && continue
+
+        # alias is assumed to not be a proper e-mail
+        # these aliases do not need to be added to Dovecot's userdb
+        [[ ! ${ALIAS} == *@* ]] && continue
```

Comment:
This problem only applies to inbound external mail AFAIK.

I would need to check if `smtpd_recipient_restrictions` apply or not when a local service in the container sends mail to Postfix to deliver to Dovecot.

Even if it does, we have some rules earlier in these restrictions such as `permit_mynetworks` which will approve accepting the mail earlier. This prevents checking the quota policy against Dovecot. That'd probably not be great usually, but when the mail originates from you as the sender, then the backscatter concern of bouncing mail back to yourself probably doesn't apply?

I am not familiar with `amavis` other than it having something to do with spam handling. If it's not a concern, and the alias maps to an account that Dovecot receives mail for, and that quota is exceeded, then Dovecot bounces the mail (_which from descriptions I read, is repeated multiple times :man_shrugging: _)

Reply:
Thank you for clarification. FYI, I use the amavis alias because of [this](https://github.com/docker-mailserver/docker-mailserver/issues/1224#issuecomment-552575494).

Reply:
As long as postmaster@example.com is not a _real_ address (in `postfix-accounts.cf`, this PR will not change / add anything as (as @polarathene has pointed out), only 1-to-1 aliases (alias -> real user) are expandend. Should all be fine here then?

Reply:
I thought the whole point of this PR is to avoid some nasty error messages? So in any case, this PR should not harm in any way. Or do I miss something fundamental at the moment?

postmaster@example.com is a real address (in postfix-accounts.cf). Why it shouldn't?

Reply:
> [postmaster@example.com](mailto:postmaster@example.com) is a real address (in postfix-accounts.cf). Why it shouldn't?

This was just an assumption (in this case, an entry will be added to Dovecot's userdb). This PR (should) does not do harm in any way, of course.

Reply:
Got it. But then, when someone have something like this in `postfix-virtual.cf`, this would not add an entry to dovecots userdb?

`amavis not-a-real-account-just-another-alias@example.com` in `postfix-virtual.cf`

PS: I don't know, if it's valid to point an alias to an other alias.

Reply:
> That's why I have something like this: `amavis postmaster@example.com` in `postfix-virtual.cf`.

I believe that should be in `postfix-aliases.cf` (_parameter `main.cf:alias_maps`_) like the comment you linked to mentions with aliasing the `amavis` account to `root` (which can alias again to `postmaster@<FQDN>`.

`postfix-virtual.cf` (_parameter `main.cf:virtual_alias_map` with generated file `/etc/postfix/virtual`_) and `postfix-accounts.cf` (_parameter `main.cf:virtual_mailbox_maps` with generated file `/etc/postfix/vmailbox`_) provide unique domains for the generated file `/etc/postfix/vhost` (_parameter `main.cf:virtual_mailbox_domains`).

Postfix should receive mail from local account without any FQDN, eg from `amavis` and match it to an alias in `hash:/etc/aliases` (_`/etc/aliases` is plain-text, hash database generated by `postalias` is `/etc/aliases.db` that Postfix uses, values are sourced from `postfix-aliases.cf` populating `/etc/aliases` at startup or change event_). Mappings in `/etc/aliases` lack an FQDN, but Postfix will under some conditions receive mail and extract the local part (eg left side of `@`) from incoming mail recipient and lookup these `/etc/aliases` if no explicit match was found in virtual aliases.

Since `postfix-aliases.cf` (_unix user aliases_) is purely meant for these aliases, while `postfix-accounts.cf` (_virtual mailboxes, storage not tied to unix user_) and `postfix-virtual.cf` (_virtual aliases, no unix user_) are handled by us differently (_contributing to vhost domains_), I think you would place `amavis` and the like in `postfix-aliases.cf` :sweat_smile: (don't worry, I'm probably as confused as you are right now :stuck_out_tongue: )

---

> Got it. But then, when someone have something like this in `postfix-virtual.cf`, this would not add an entry to dovecots userdb?

[Postfix address rewriting docs on virtual aliases](http://www.postfix.org/ADDRESS_REWRITING_README.html#virtual) shows an example of a 1-to-1 mapping in `postfix-virtual.cf` that's got no associated FQDN, so even I'm confused about it vs the `postfix-alias.cf` kind. This is perfectly valid as you point out..

Dovecot AFAIK doesn't care what the user name is, eg `amavis` or `amavis@example.com`? What I believe happens here is Postfix will append `main.cf:myorigin` or if not available another fallback domain it's aware of, and then deliver that to Dovecot to store in `/var/mail`.

In the case of Amavis, it should only be mail to and from your server, if Dovecot can't be configured to ignore quota for system user accounts (eg: `postmaster@example.com`) you could make the quota really high. In regards to the issue this PR is solving, there are "permit" rules that can come before the quota policy check, eg `mynetworks` and the one for SASL authenticated is probably safe and both of these are configured to come before the quota check, thus Postfix will accept and if Dovecot rejects, it bounces but if the accounts that sent it aren't external, the backscatter blacklisting concern shouldn't apply?

---

> PS: I don't know, if it's valid to point an alias to an other alias.

I have read of an `all` alias that expands to many users, some which I guess could be aliases. 1-to-1 alias that maps to another alias seems a bit silly, but our docs have an example of working around `SPOOF_PROTECTION` issue for [aliasing to gmail via an alias to alias mapping which then maps the 2nd level alias to the external account](https://github.com/docker-mailserver/docker-mailserver/blob/fb72f3ad52769ad470cd1f7fc36f76b53853f3e8/docs/content/examples/tutorials/basic-installation.md?plain=1#L130-L135).

---

Probably good to eventually investigate further the above for definitive answers about account/alias handling and document that somewhere, but I don't have time to allocate to that right now :grimacing:

Reply:
Just to clarify, the log issue with Postfix asking Dovecot for quota status of an alias that Dovecot doesn't know about is always the `RCPT TO` (envelope recipient) address the email was originally sent to.

Regardless of nested chains, Postfix does not expand that aliases at this point (_which could be N levels deep or expand to multiple addresses to send to_). In the docs mention in my previous comment, that means the gmail account with double aliases won't be added to Dovecot, not that it matters since that's aliasing to an external account anyway.

If it ever turns out to be an issue, we can probably regard it as niche and unsupported. Our responsibility would just be to make the caveat with quota support with aliases visible in our docs, and that we only assist with direct aliasing to a single real mailbox account, any fancy aliasing beyond that puts the user as risk of backscatter if they have quotas enabled.

Reply:
> don't worry, I'm probably as confused as you are right now 

😆 

Thank you for you detailed answer!